src/databricks/labs/remorph/transpiler/lsp/lsp_engine.py

```diff
@@ -35,12 +68,175 @@ def parse(cls, data: dict[str, Any]) -> _LSPRemorphConfigV1:
+class TranspileDocumentResponse:
+    # 'id' is mandated by LSP
+    # pylint: disable=invalid-name
+    id: int | str = attrs.field()
+    result: TranspileDocumentResult = attrs.field()
+    jsonrpc: str = attrs.field(default="2.0")
+
+
+_SPECIAL_PROPERTIES.extend(
+    [f"{TranspileDocumentRequest.__name__}.method", f"{TranspileDocumentRequest.__name__}.jsonrpc"]
+)
+METHOD_TO_TYPES[TRANSPILE_TO_DATABRICKS_METHOD] = (
+    TranspileDocumentRequest,
+    TranspileDocumentResponse,
+    TranspileDocumentParams,
+    None,
+)
+
+
+# subclass BaseLanguageClient so we can override stuff when required
+class _LanguageClient(BaseLanguageClient):
+
+    def __init__(self):
+        version = ProductInfo.from_class(type(self)).version()
+        super().__init__("Remorph", version)
+        self._transpile_to_databricks_capability: Registration | None = None
+        self._register_lsp_features()
+
+    @property
+    def is_alive(self):
+        return self._server and self._server.returncode is None
+
+    @property
+    def transpile_to_databricks_capability(self):
+        return self._transpile_to_databricks_capability
+
+    @lsp_feature(CLIENT_REGISTER_CAPABILITY)
+    def register_capabilities(self, params: RegistrationParams) -> None:
+        for registration in params.registrations:
+            if registration.method == TRANSPILE_TO_DATABRICKS_METHOD:
+                logger.debug(f"Registered capability: {registration.method}")
+                self._transpile_to_databricks_capability = registration
+                continue
+            logger.debug(f"Unknown capability: {registration.method}")
+
+    async def transpile_document_async(self, params: TranspileDocumentParams) -> TranspileDocumentResult:
+        if self.stopped:
+            raise RuntimeError("Client has been stopped.")
+        await self._await_for_transpile_capability()
+        return await self.protocol.send_request_async(TRANSPILE_TO_DATABRICKS_METHOD, params)
+
+    async def _await_for_transpile_capability(self):
+        for _ in range(1, 10):
+            if self.transpile_to_databricks_capability:
+                return
+            await asyncio.sleep(0.1)
+        if not self.transpile_to_databricks_capability:
+            raise FeatureRequestError(f"LSP server did not register its {TRANSPILE_TO_DATABRICKS_METHOD} capability")
+
+    # can't use @client.feature because it requires a global instance
+    def _register_lsp_features(self):
+        for name, options, func in _LSP_FEATURES:
+            decorator = self.protocol.fm.feature(name, options)
+            wrapper = self._wrapper_for_lsp_feature(func)
+            decorator(wrapper)
+
+    def _wrapper_for_lsp_feature(self, func):
+        def wrapper(params):
+            return func(self, params)
+
+        return wrapper
+
+
+class ChangeManager(abc.ABC):
+
+    @classmethod
+    def apply(cls, source_code: str, changes: Sequence[TextEdit]) -> str:
+        lines = source_code.split("\n")
```

Comment:
@ericvergnaud  How will a SQL procedure, or any complex structure, be managed using this change manager?

Wouldn't the transpiler want the entire stream of syntax be parsed or analyzed to be able to make changes.

Reply:
The transpiler sits behind the protocol not within (the LSP protocol deals with text, not with complex structures).
The way it works is as follows:
  - client sends full text of a document to the server
  - client invokes a server capability or command
  - server returns a list of text edits that should be applied (this is a mandatory LSP optimization where only changes are sent, not the entire text)
  - client applies the changes
 The purpose of this class is to apply the changes.